README.md
[![npm](https://img.shields.io/npm/v/homebridge-config-ui-rdp.svg)](https://www.npmjs.com/package/homebridge-config-ui-rdp) 
[![npm](https://img.shields.io/npm/dt/homebridge-config-ui-rdp.svg)](https://www.npmjs.com/package/homebridge-config-ui-rdp)

:gb: :de: :fr: :netherlands: :belgium: :poland: :czech_republic: :ru:

# Homebridge Config UI RDP

This is a plugin for [Homebridge](https://github.com/nfarina/homebridge). This is a fork of the work originally done by [oznu](https://github.com/oznu/homebridge-config-ui-x).

This plugin allows you to monitor, backup and configure your Homebridge server from a browser.

## Installation Instructions

```
sudo npm install -g --unsafe-perm homebridge-config-ui-rdp
```

Once you have installed and configured the plugin you can access the interface via http://localhost:8080. 

The default username is `admin` and the default password is `admin`.

# Configuration

Add this to your homebridge `config.json` file

```json
"platforms": [
    {
      "platform": "config",
      "name": "Config",
      "port": 8080,
      "sudo": false
    }
]
```

**Optional Settings**

* `log` - [See below for details](#log-viewer-configuration).
* `sudo` - [See below for details](#sudo-mode).
* `restart` - The command to run when a restart request is sent from the browser. If not populated it will just terminate the Homebridge process and let your process manager (like systemd) restart it.
* `temp` - The path to the file that can display your current CPU temperature. eg. `/sys/class/thermal/thermal_zone0/temp`
* `theme` - The colour scheme to use. Possible values: `red`, `pink`, `purple`, `indigo`, `blue`, `blue-grey`, `green`, `orange`. Defaults to `red`.
* `ssl` - [See below for details](#enabling-ssl)

## Accessory Control

The plugin allows you to view and control some types of Homebridge accessories from your web browser. 

To [enable accessory control](https://github.com/homebridge-config-ui-rdp/homebridge-config-ui-rdp/wiki/Enabling-Accessory-Control) you must be running Homebridge in insecure mode:

```
homebridge -I
```

*Insecure mode, as it's name suggests, is not secure. Anyone with network access to your Homebridge server port and pin will be able to control your accessories. Unless exposing the Homebridge port to the internet, this will generally not be a problem as access will be limited to people already on your local network. Remote access via a home hub (Apple TV, iPad, HomePod etc.) will still be secure. Enable at your own risk.*

Not all accessory types are supported. See [this issue](https://github.com/homebridge-config-ui-rdp/homebridge-config-ui-rdp/issues/47) for a full list of supported accessory types.

## Log Viewer Configuration

Homebridge Config UI X allows you to view the homebridge process logs in the browser. These logs can be loaded from a file or from a command.

### Logs From File

Example loading logs from a file, change `/var/log/homebridge.log` to the actual location of your log file:

```json
"platforms": [
    {
      "platform": "config",
      "name": "Config",
      "port": 8080,
      "log": {
        "method": "file",
        "path": "/var/log/homebridge.log"
      }
    }
]
```

*Make sure the user which is running the Homebridge process has the correct permissions to read the log file. You may need to enable the [sudo option](#sudo-mode) to avoid permission errors if you are not running Homebridge as root.*

### Logs From Systemd

If you're using `systemd` to manage the Homebridge process then you can just set the `method` to `systemd`:

```json
"platforms": [
    {
      "platform": "config",
      "name": "Config",
      "port": 8080,
      "restart": "sudo -n systemctl restart homebridge",
      "log": {
        "method": "systemd",
        "service": "homebridge"
      }
    }
]
```

*You may need to enable the [sudo option](#sudo-mode) to avoid permission errors if you are not running Homebridge as root.*

### Logs From Custom Command

The `log` option can alternatively specify a command to spawn that will stream the logs to the client. This command should stream the logs to `stdout`:

```json
"platforms": [
    {
      "platform": "config",
      "name": "Config",
      "port": 8080,
      "log": {
        "method": "custom",
        "command": "sudo -n tail -n 100 -f /var/log/homebridge.log"
      }
    }
]
```

## Sudo Mode

Many operations performed by Homebridge Config UI X, such as installing plugins, upgrading Homebridge and viewing the logs can require root permissions. You can run the Homebridge service as root or you can enable the `sudo` option in the config.

```json
"platforms": [
    {
      "platform": "config",
      "name": "Config",
      "port": 8080,
      "sudo": true
    }
]
```

When `sudo` mode is enabled Homebridge Config UI X will use `sudo` when executing installing, removing or upgrading plugins, viewing the logs using the [Logs From File](#logs-from-file) or [Logs From Systemd](#logs-from-systemd) method, and when upgrading Homebridge. It will not be used for [Logs From Custom Command](#logs-from-custom-command) or custom restart commands.

### Password-less sudo required

For `sudo` mode to work password-less sudo is required. You can enable password-less sudo by adding this entry to the bottom of your `/etc/sudoers` file (use `visudo` to edit the file!):

```
homebridge    ALL=(ALL) NOPASSWD: ALL
```

*Replace `homebridge` with the actual user you are running Homebridge as.*

## Enabling SSL

You can run this plugin over an encrypted HTTPS connection by configuring the `ssl` options.

```json
"platforms": [
    {
      "platform": "config",
      "name": "Config",
      "port": 8080,
      "ssl": {
        "key": "/path/to/privkey.pem",
        "cert": "/path/to/fullchain.pem"
      }
    }
]
```

Or if using a **PKCS#12** certificate you can setup SSL like this:

```json
"platforms": [
    {
      "platform": "config",
      "name": "Config",
      "port": 8080,
      "ssl": {
        "pfx": "/path/to/cert.pfx",
        "passphrase": "sample"
      }
    }
]
```

# Usage

### Status Screen

This shows you that the services are running. It also has your HomeKit pin.

![Status](screenshots/homebridge-config-ui-rdp-status.png)

### Log Screen

This shows you the rolling log. This is helpful for troubleshooting.

![Log](screenshots/homebridge-config-ui-rdp-logs.png)

### Plugin Screen

This shows you the currently installed plugins and allows you to install, remove and upgrade plugins.

![Log](screenshots/homebridge-config-ui-rdp-plugins.png)

When installing, removing or upgrading plugins the output is show in the browser to help troubleshoot any issues.

![Log](screenshots/homebridge-config-ui-rdp-plugins-install.png)


### Configuration Screen

And finally the configuration screen allows you to modify your Homebridge settings and your platforms and accessories.

![Config](screenshots/homebridge-config-ui-rdp-config.png)

# Supported Browsers

The following browsers are supported by this plugin:

* Safari >= 10
* Chrome >= 55 (including Opera)
* FireFox >= 58
* MS Edge >= 13 on the desktop
* iOS >=10 and Chrome on mobile

MS Internet Explorer (any version) is not supported!

# Contributing

Please see [CONTRIBUTING.md](CONTRIBUTING.md).

# Troubleshooting

#### 1. Errors during installation

Make sure you installed the package with `sudo` and used the  `--unsafe-perm` flag. Most installation errors can be fixed by removing the plugin and reinstalling:

```shell
# cleanup
sudo npm uninstall -g homebridge-config-ui-rdp

# reinstall
sudo npm install -g --unsafe-perm homebridge-config-ui-rdp
```

#### 2. Accessories tab missing

If the Accessories tab is not show then you are not running Homebridge in insecure mode. See the [Enabling Accessory Control](https://github.com/homebridge-config-ui-rdp/homebridge-config-ui-rdp/wiki/Enabling-Accessory-Control) wiki for details. If you have just enabled insecure mode make sure you have restarted Homebridge and refreshed the page in your browser.

#### 3. Ask on Slack

[![Slack Status](https://slackin-znyruquwmv.now.sh/badge.svg)](https://slackin-znyruquwmv.now.sh)

Join the [Homebridge Slack](https://slackin-znyruquwmv.now.sh/) chat and ask in the [#ui](https://homebridgeteam.slack.com/messages/C9NH0CUTY) channel.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This project uses [Semantic Versioning](https://semver.org/).

## 4.0.3 (2018-08-27)

* Updated to new version

## 3.8.9 (2018-08-06)

### Bug Fixes

* **i18n:** Improve Russian Translations ([#146](https://github.com/RienduPre/homebridge-config-ui-rdp/pull/146))

### Other Changes

* Updated Angular to 6.1.1
* Updated npm dependencies

## 3.8.8 (2018-07-30)

### Notable Changes

* **i18n:** Czech language translation added by [@HonzaaD](https://github.com/HonzaaD)
* **i18n:** Russian language translation added by [@Caribsky](https://github.com/Caribsky)
  * User contributions to the non-english translations are always welcome, [click here](https://github.com/RienduPre/homebridge-config-ui-rdp/blob/master/CONTRIBUTING.md#contributing-to-translations) for details on how you can help

### Bug Fixes

* **Config Editor:** Backup button actually downloads up the `config.json` now ([#144](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/144))
* **Config Editor:** Fixed issued using config editor on mobile devices ([#131](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/131))
* **Accessory Control:** Accessory layout changes are now persistent again

## 3.8.7 (2018-07-28)

### Notable Changes

* **Accessory Control:** Long-clicking a lightbulb with no additional characteristics (eg. brightness) no longer opens a modal ([#47#issuecomment-405089113](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/47#issuecomment-405089113))
* **Accessory Control:** Right-clicking an accessory on a non-mobile device will bring up a modal showing all accessory characteristics and other information
* **i18n:** Polish language translation added by [@mientki](https://github.com/mientki)
  * User contributions to the non-english translations are always welcome, [click here](https://github.com/RienduPre/homebridge-config-ui-rdp/blob/master/CONTRIBUTING.md#contributing-to-translations) for details on how you can help

## 3.8.6 (2018-07-27)

### Bug Fixes

* **Config Editor:** Backup button more reliable ([#135](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/135))
* **Accessory Control:** Refresh all accessories when one is changed to ensure the dashboard is up-to-date ([#136](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/136))

### Other Changes

* Updated Angular to 6.1.0
* Updated npm dependencies

## 3.8.5 (2018-06-15)

### Bug Fixes

* Handle loading config schema for [@homebridge](https://www.npmjs.com/org/homebridge) plugins correctly

## 3.8.4 (2018-06-14)

### Notable Changes

* Made `config.json` saves a bit more robust in an attempt to fix ([#122](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/122))
* Added `proxyHost` config option to make running behind a reverse proxy easier ([#119](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/119))

## 3.8.3 (2018-06-07)

### Notable Changes

* Fixes to German Translation ([#116](https://github.com/RienduPre/homebridge-config-ui-rdp/pull/116))
* Allow web terminal access on Linux-based hosts *only* if a certain environment variable is set ([#110](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/110))
  * Homebridge must be running with the `HOMEBRIDGE_CONFIG_UI_TERMINAL=1` environment variable to access the terminal

## 3.8.2 (2018-06-06)

### Other Changes

* Minor bug fixes
* Updated npm dependencies

## 3.8.0 (2018-06-03)

### Notable Changes

* Added support for non-english translations ([#102](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/102))
  * German translation added with the assistance of [@razer4908](https://github.com/razer4908)
  * French translation added by [@the0neyouseek](https://github.com/the0neyouseek) ([#105](https://github.com/RienduPre/homebridge-config-ui-rdp/pull/105))
  * Language is selected automatically based on browser settings
  * If you want to contribute to the translation work please see [CONTRIBUTING.md](https://github.com/RienduPre/homebridge-config-ui-rdp/blob/master/CONTRIBUTING.md#contributing-to-translations)
* Imperial display units now apply to temperature sensors and thermostat accessories ([#96](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/96))

### Other Changes

* Updated npm dependencies
* Config changes for the UI should require a full page refresh less often

## 3.7.0 (2018-05-31)

### Notable Changes

* Added support for imperial temperature display units ([#96](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/96))

## 3.6.6 (2018-05-26)

### Notable Changes

* **ui**: upgraded from Angular 5.x to 6.x
* updated npm dependencies

## 3.6.5 (2018-05-20)

### Notable Changes

* Updated npm dependencies

## 3.6.3 (2018-05-10)

### Notable Changes

* Removed the `nsp` package as the tool has been discontinued, vulnerability scanning is being added to `npm` natively
* Updated npm dependencies

## 3.6.2 (2018-05-09)

### Other Changes

* Updated npm dependencies
* Errors will no longer be thrown if a plugin is installed with CVSS alerts lower than 4 (Low Risk)

## 3.6.1 (2018-05-08)

### Bug Fixes

* Fixed bug that could cause plugins that take a long time to install or upgrade to fail

## 3.6.0 (2018-05-08)

### Notable Changes

* **Beta Feature**: Initial implementation of GUI/form based setup for supported plugins
* Added support for optional native HTTPS / SSL ([#68](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/68), [#35](https://github.com/RienduPre/homebridge-config-ui-rdp/issues/35))
* The Log Viewer config options have changed, existing options have been have depreciated, see [README](https://github.com/RienduPre/homebridge-config-ui-rdp#log-viewer-configuration) for details
* Docker users may now configure this plugin using the `config.json` or the new plugin GUI/form config method

src/hb.ts
import * as os from 'os';
import * as path from 'path';
import * as fs from 'fs-extra';
import * as color from 'bash-color';
import * as commander from 'commander';
import * as semver from 'semver';

import { WSS } from './wss';

class HomebridgeUI {
  public ui: any;
  public pluginName: string;
  public homebridgeVersion: string;
  public homebridgeInsecure: boolean;
  public homebridgeNpmPkg: string;
  public homebridgeFork: string;
  public homebridgeConfig: HomebridgeConfigType;
  public runningInDocker: boolean;
  public runningInLinux: boolean;
  public linuxServerOpts: { restart?: string; shutdown?: string; };
  public enableTerminalAccess: boolean;
  public ableToConfigureSelf: boolean;
  public configPath: string;
  public authPath: string;
  public storagePath: string;
  public pluginPath: string;
  public port: number | string;
  public proxyHost: string;
  public logOpts: {
    method: 'file' | 'systemd' | 'custom';
    path?: string;
    service?: string;
    command?: string;
    tail?: string; /** @deprecated since 5.6.0 */
    systemd?: string; /** @deprecated since 5.6.0 */
  } | string;
  public restartCmd;
  public useSudo: boolean;
  public authMethod: string | boolean;
  public formAuth: boolean;
  public theme: string;
  public loginWallpaper: string;
  public availableThemes: string[];
  public temperatureFile: string;
  public temperatureUnits: 'c' | 'f';
  public accessoryLayoutPath: string;
  public wss: WSS;

  constructor() {
    this.ui = fs.readJSONSync(path.resolve(__dirname, '../package.json'));

    this.availableThemes = [
      'red',
      'pink',
      'purple',
      'indigo',
      'blue',
      'blue-grey',
      'green',
      'orange'
    ];
  }

  public init(setup) {
    this.configPath = setup.configPath;
    this.authPath = path.join(setup.storagePath, 'auth.json');
    this.storagePath = setup.storagePath;
    this.accessoryLayoutPath = path.resolve(setup.storagePath, 'accessories', 'uiAccessoriesLayout.json');
    this.homebridgeVersion = setup.homebridgeVersion;

    this.parseConfig(setup.config);
    this.parseCommandLineArgs();
  }

  private parseConfig(config) {
    this.pluginName = config.name || this.ui.name;
    this.port = config.port || 8080;
    this.proxyHost = config.proxyHost;
    this.logOpts = config.log;
    this.restartCmd = config.restart;
    this.useSudo = config.sudo;
    this.authMethod = config.auth;
    this.homebridgeFork = config.fork;
    this.homebridgeNpmPkg = config.homebridgeNpmPkg || 'homebridge';
    this.homebridgeInsecure = config.homebridgeInsecure;
    this.pluginPath = config.pluginPath;
    this.runningInDocker = Boolean(process.env.HOMEBRIDGE_CONFIG_UI === '1');
    this.runningInLinux = (!this.runningInDocker && os.platform() === 'linux');
    this.linuxServerOpts = config.linux || {};
    this.enableTerminalAccess = this.runningInDocker || Boolean(process.env.HOMEBRIDGE_CONFIG_UI_TERMINAL === '1');
    this.ableToConfigureSelf = (!this.runningInDocker || semver.satisfies(process.env.CONFIG_UI_VERSION, '>=3.5.5'));
    this.loginWallpaper = config.loginWallpaper;
    this.temperatureUnits = config.tempUnits || 'c';

    if (config.auth === 'none' || config.auth === false) {
      this.formAuth = false;
    } else if (config.auth === 'basic') {
      this.formAuth = false;
    } else {
      this.formAuth = true;
    }

    // check theme is valid
    if (config.theme && this.availableThemes.find(x => x === config.theme)) {
      this.theme = config.theme;
    } else if (config.theme) {
      // delay the output of the warning message so it does not get lost under homebridge setup details
      setTimeout(() => {
        this.warn(`Invalid theme in config.json. Possible options are: ${this.availableThemes.join(', ')}`);
      }, 2000);
      this.theme = 'red';
    } else {
      this.theme = 'red';
    }

    // check the path to the temp file actually exists
    if (config.temp && fs.existsSync(config.temp)) {
      this.temperatureFile = config.temp;
    } else if (config.temp) {
      // delay the output of the warning message so it does not get lost under homebridge setup details
      setTimeout(() => {
        this.warn(`WARNING: Configured path to temp file does not exist: ${config.temp}`);
        this.warn(`WARNING: CPU Temp will not be displayed`);
      }, 2000);
    }
  }

  private parseCommandLineArgs() {
    // parse plugin path argument from homebridge
    commander
      .allowUnknownOption()
      .option('-P, --plugin-path [path]', '', (p) => this.pluginPath = p)
      .option('-I, --insecure', '', () => this.homebridgeInsecure = true)
      .parse(process.argv);
  }

  public async refreshHomebridgeConfig() {
    try {
      this.homebridgeConfig = await import(this.configPath);
    } catch (e) {
      this.homebridgeConfig = {
        bridge: {
          name: 'Homebridge',
          port: 51826
        }
      };
      this.error(`Failed to load ${this.configPath} - ${e.message}`);
    }

    if (!this.homebridgeConfig.bridge.port) {
      this.error('Homebridge config.json error: bridge.port missing.');
    }
  }

  public async updateConfig(config) {
    const now = new Date();

    if (!config) {
      config = {};
    }

    if (!config.bridge) {
      config.bridge = {};
    }

    if (!config.bridge.name) {
      config.bridge.name = 'Homebridge';
    }

    if (!config.bridge.port) {
      config.bridge.port = 51826;
    }

    if (!config.bridge.username) {
      config.bridge.username = this.generateUsername();
    }

    if (!config.bridge.pin) {
      config.bridge.pin = this.generatePin();
    }

    if (!config.accessories) {
      config.accessories = [];
    }

    if (!config.platforms) {
      config.platforms = [];
    }

    // create backup of existing config
    await fs.rename(this.configPath, `${this.configPath}.${now.getTime()}`);

    // save config file
    fs.writeJsonSync(this.configPath, config, { spaces: 4 });

    this.log('Changes to config.json saved.');

    return config;
  }

  public async listConfigBackups() {
    const dirContents = await fs.readdir(this.storagePath);

    const backups = dirContents
      .filter(x => x.indexOf('config.json.') === 0)
      .sort()
      .reverse()
      .map(x => {
        const ext = x.split('.');
        if (ext.length === 3 && !isNaN(ext[2] as any)) {
          return {
            id: ext[2],
            timestamp: new Date(parseInt(ext[2], 10)),
            file: x
          };
        } else {
          return null;
        }
      })
      .filter((x => x && !isNaN(x.timestamp.getTime())));

    return backups;
  }

  public async getConfigBackup(backupId: string) {
    // check backup file exists
    if (!fs.existsSync(this.configPath + '.' + parseInt(backupId, 10))) {
      throw new Error(`Backup ${backupId} Not Found`);
    }

    // read source backup
    return await fs.readFile(this.configPath + '.' + parseInt(backupId, 10));
  }

  public async deleteAllConfigBackups() {
    const backups = await this.listConfigBackups();

    // delete each backup file
    await backups.forEach(async(backupFile) => {
      await fs.unlink(path.resolve(this.storagePath, backupFile.file));
    });
  }

  public async resetHomebridgeAccessory() {
    // // load config file
    // const config: HomebridgeConfigType = await fs.readJson(this.configPath);

    // // generate new random username and pin
    // if (config.bridge) {
    //   config.bridge.pin = this.generatePin();
    //   config.bridge.username = this.generateUsername();

    //   this.log(`Homebridge Reset: New Username: ${config.bridge.username}`);
    //   this.log(`Homebridge Reset: New Pin: ${config.bridge.pin}`);

    //   // save config file
    //   await this.updateConfig(config);
    // } else {
    //   this.error('Homebridge Reset: Could not reset homebridge username or pin. Config format invalid.');
    // }

    // remove accessories and persist directories
    await fs.remove(path.resolve(this.storagePath, 'accessories'));
    await fs.remove(path.resolve(this.storagePath, 'persist'));

    this.log(`Homebridge Reset: "persist" directory removed.`);
    this.log(`Homebridge Reset: "accessories" directory removed.`);
  }

  private generatePin() {
    let code: string | Array<any> = Math.floor(10000000 + Math.random() * 90000000) + '';
    code = code.split('');
    code.splice(3, 0, '-');
    code.splice(6, 0, '-');
    code = code.join('');
    return code;
  }

  private generateUsername() {
    const hexDigits = '0123456789ABCDEF';
    let username = '0E:';
    for (let i = 0; i < 5; i++) {
      username += hexDigits.charAt(Math.round(Math.random() * 15));
      username += hexDigits.charAt(Math.round(Math.random() * 15));
      if (i !== 4) { username += ':'; }
    }
    return username;
  }

  public async getAccessoryLayout(user) {
    if (fs.existsSync(this.accessoryLayoutPath)) {
      const accessoryLayout = await fs.readJson(this.accessoryLayoutPath);
      if (user in accessoryLayout) {
        return accessoryLayout[user];
      }
    }
    return [
      {
        name: 'Default Room',
        services: []
      }
    ];
  }

  public async updateAccessoryLayout(user, layout) {
    let accessoryLayout;

    try {
      accessoryLayout = await fs.readJson(this.accessoryLayoutPath);
    } catch (e) {
      accessoryLayout = {};
    }

    accessoryLayout[user] = layout;
    fs.writeJsonSync(this.accessoryLayoutPath, accessoryLayout);
    this.log(`[${user}] Accessory layout changes saved.`);
    return layout;
  }

  public log(...params) {
    console.log(
      color.white(`[${new Date().toLocaleString()}]`), color.cyan(`[${this.pluginName}]`), ...params
    );
  }

  public warn(...params) {
    console.warn(
      color.white(`[${new Date().toLocaleString()}]`), color.cyan(`[${this.pluginName}]`), color.yellow(params.join(' '))
    );
  }

  public error(...params) {
    console.error(
      color.white(`[${new Date().toLocaleString()}]`), color.cyan(`[${this.pluginName}]`), color.red(params.join(' '))
    );
  }
}

export interface HomebridgeConfigType {
  bridge: {
    name: string;
    username?: string;
    port: number;
    pin?: string;
  };
  platforms?: {
    name: string;
    platform: string;
  }[];
  accessories?: {
    accessory: string;
    name: string;
  }[];
}

export const hb = new HomebridgeUI();
